File: lua/protocol.lua
local memory_handler = require('scripts/memory')
local util = require('scripts/util')

local M = {}

M.observations_to_send = 0

local SendType = {
  OBSERVATION_START = 0,
  SCREENSHOT = 1,
  MATCH_STATUS = 2,
  GAME_PAUSED = 3,
  PLAYER1_HP = 4,
  PLAYER2_HP = 5,
  OBSERVATION_END = 255
}

local ParseType = {
  ACTION = 1,
  RESET = 2,
  OBSERVATIONS_REQUEST = 3,
  CLEAR_CONTROLLER = 4
}

function M:parse(connection, opcode)
  if opcode == ParseType.ACTION then
    self:parse_action(connection)
  elseif opcode == ParseType.RESET then
    self:parse_reset(connection)
  elseif opcode == ParseType.OBSERVATIONS_REQUEST then
    self:parse_observations_request(connection)
  elseif opcode == ParseType.CLEAR_CONTROLLER then
    self:parse_clear_controller(connection)
  else
    print("Invalid opcode: ", opcode)
  end
end

function M:parse_clear_controller(connection)
  local slot_id = connection.client:readU8()
  util.clear_keys(slot_id)
end

function M:parse_reset(connection)
  local state = connection.client:readU8()

  util.clear_keys(1)
  util.clear_keys(2)
  util.load_state(state)
end

function M:parse_action(connection)
  local slot_id = connection.client:readU8()
  local first_key = connection.client:readU8()
  local second_key = connection.client:readU8()

  util.clear_keys(slot_id)
  util.set_keys_cb = function ()
    util.set_key(slot_id, first_key)
    util.set_key(slot_id, second_key)
  end
  util.set_keys_cb_countdown = 2
end

function M:parse_observations_request(connection)
  M.observations_to_send = connection.client:readU8()
end

function M:send_observation(connection)
  if not connection.connected then
    print("Tried to send observation but client is not connected.")
    return
  end

  connection.client:writeU8(SendType.OBSERVATION_START)
  self:send_screenshot(connection)
  self:send_match_status(connection)
  self:send_game_paused(connection)
  self:send_player1_hp(connection)
  self:send_player2_hp(connection)
  connection.client:writeU8(SendType.OBSERVATION_END)
end

function M:send_match_status(connection)
  connection.client:writeU8(SendType.MATCH_STATUS)
  local match_status, _ = memory_handler.read_from_memory(memory_handler.address.match_status)
  connection.client:writeU8(match_status)
end

function M:send_game_paused(connection)
  connection.client:writeU8(SendType.GAME_PAUSED)
  local game_paused, _ = memory_handler.read_from_memory(memory_handler.address.game_paused)
  connection.client:writeU8(game_paused)
end

function M:send_player1_hp(connection)
  connection.client:writeU8(SendType.PLAYER1_HP)
  local player1_hp, _ = memory_handler.read_from_memory(memory_handler.address.player1_hp)
  connection.client:writeU8(player1_hp)
end

function M:send_player2_hp(connection)
  connection.client:writeU8(SendType.PLAYER2_HP)
  local player2_hp, _ = memory_handler.read_from_memory(memory_handler.address.player2_hp)
  connection.client:writeU8(player2_hp)
end

function M:send_screenshot(connection)
  connection.client:writeU8(SendType.SCREENSHOT)
  local screenshot = PCSX.GPU.takeScreenShot()
  connection.client:writeI32(screenshot.height)
  connection.client:writeI32(screenshot.width)

  local bpp = 16
  if tonumber(screenshot.bpp) == 1 then
    bpp = 24
  end
  connection.client:writeU8(bpp)

  screenshot.data = tostring(screenshot.data)
  connection.client:writeU32(#screenshot.data)
  connection.client:write(screenshot.data)
end

return M

File: python/protocol.py
from observation import * 
from agent import *
from enum import IntEnum
from PIL import Image
import struct

OPCODE_SIZE = 1 # in bytes

class ParseType(IntEnum):
  OBSERVATION_START = 0
  SCREENSHOT = 1
  MATCH_STATUS = 2
  GAME_PAUSED = 3
  PLAYER1_HP = 4
  PLAYER2_HP = 5
  OBSERVATION_END = 255 

class SendType(IntEnum):
  ACTION = 1
  RESET = 2
  OBSERVATIONS_REQUEST = 3
  CLEAR_CONTROLLER = 4

class NetworkParser:
  def __init__(self, connection):
    self.connection = connection

  def parse_observation(self):
    assert self.readU8() == ParseType.OBSERVATION_START
    observation = Observation()
    while (opcode := self.readU8()) != ParseType.OBSERVATION_END:
      match opcode:
        case ParseType.SCREENSHOT:
          observation.frame = self.parse_screenshot()
        case ParseType.MATCH_STATUS:
          observation.match_status = self.parse_match_status()
        case ParseType.GAME_PAUSED:
          observation.game_paused = self.parse_game_paused()
        case ParseType.PLAYER1_HP:
          observation.player1_hp = self.parse_player1_hp()
        case ParseType.PLAYER2_HP:
          observation.player2_hp = self.parse_player2_hp()
        case _:
          raise ValueError(f"Received invalid opcode: {opcode}")
    return observation

  def parse_match_status(self):
    match_status = self.readU8() 
    # 0 -> not in match
    # 1 -> in match
    # 2 -> match finished
    assert match_status >= 0 and match_status <= 2, f"Parsed invalid match_status, its equal to {match_status}"
    return match_status

  def parse_game_paused(self):
    game_paused = self.readU8() 
    # 0 -> not paused
    # 1 -> paused
    assert game_paused == 0 or game_paused == 1, f"Parsed invalid game_paused, its equal to {game_paused}"
    return game_paused 
  
  def parse_player1_hp(self):
    player1_hp = self.readU8() 
    # player1 max hp is 200
    # assert player1_hp >= 0 and player1_hp <= 200, f"Parsed invalid player1_hp, its equal to {player1_hp}"
    return player1_hp 

  def parse_player2_hp(self):
    player2_hp = self.readU8() 
    # player2 max hp is 200 
    # assert player2_hp >= 0 and player2_hp <= 200, f"Parsed invalid player2_hp, its equal to {player2_hp}"
    return player2_hp 

  def parse_screenshot(self):
    height = self.readI32() 
    assert height > 0, f"Parsed invalid height from screenshot, its equal to {height}"

    width = self.readI32()
    assert width > 0, f"Parsed invalid width from screenshot, its equal to {width}"

    bpp = self.readU8() 
    assert bpp == 16 or bpp == 24, f"Parsed invalid bpp from screenshot, its equal to {bpp}"

    screenshot_len = self.readI32() 
    screenshot = self.read_raw(screenshot_len)

    mode = "RGB" if bpp == 24 else "RGB;15"
    return Image.frombytes("RGB", (width, height), bytes(screenshot), 'raw', mode)

  def send_action(self, action, player):
    # first byte opcode
    # second byte 1 or 2 indicating player 1 or 2
    # third and fourth byte action
    assert player == 1 or player == 2, f"Trying to send invalid player, its equal to {player}"
    packet = struct.pack('!BBBB', SendType.ACTION, player, *action)
    self.connection.sendall(packet)

  def send_reset(self, state):
    packet = struct.pack('!BB', SendType.RESET, state)
    self.connection.sendall(packet)

  def send_observations_request(self, num_of_observations):
    packet = struct.pack('!BB', SendType.OBSERVATIONS_REQUEST, num_of_observations)
    self.connection.sendall(packet)
    
  def send_clear_controller(self, player):
    packet = struct.pack('!BB', SendType.CLEAR_CONTROLLER, player)
    self.connection.sendall(packet)

  def readU8(self):
    return ord(self.connection.recv(OPCODE_SIZE))

  def readI32(self):
    return int.from_bytes(self.connection.recv(4), byteorder="little")

  def read_raw(self, length):
    data = bytearray()
    while len(data) < length:
      chunk = self.connection.recv(length - len(data))
      data.extend(chunk)
    return data
